sys/pm_layered/pm.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_periph_pm
+ * @{
+ *
+ * @file
+ * @brief       Platform-independent power management code
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include "irq.h"
+#include "periph/pm.h"
+#include "pm_layered.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#ifndef PM_NUM_MODES
+#error PM_NUM_MODES must be defined in periph_cpu.h!
+#endif
+
+#ifndef PM_BLOCKER_INITIAL
+#define PM_BLOCKER_INITIAL { .val_u32=0x01010101 }
+#endif
+
+/**
+ * @brief Power Management mode typedef
+ */
+typedef union {
+    uint32_t val_u32;
+    uint8_t val_u8[PM_NUM_MODES];
+} pm_blocker_t;
+
+/**
+ * @brief Global variable for keeping track of blocked modes
+ */
+volatile pm_blocker_t pm_blocker = PM_BLOCKER_INITIAL;
+
+void pm_set_lowest(void)
+{
+    pm_blocker_t blocker = (pm_blocker_t) pm_blocker;
+    unsigned mode = PM_NUM_MODES;
+    while (mode) {
+        if (blocker.val_u8[mode-1]) {
+            break;
+        }
+        mode--;
+    }
+
+    /* set lowest mode if blocker is still the same */
+    unsigned state = irq_disable();
+    if (blocker.val_u32 == pm_blocker.val_u32) {
```

Comment:
TL;DR: I missed the previous comment about the pm_layered, sorry. Now it makes sense. Ignore my comment. Thanks!

What I meant is: first you set `blocker` as `pm_blocker`, compute into which mode to go and then when the irq is disabled, it is verified that `blocker` and `pm_blocker` are still the same (safety check). Why not doing everything into the irq_disable if at most 4 iterations will occur? I assume: keep things within irq_disable blocks short :-)

In any case, I asked because the only reason I see to have 4 power modes is because of the `val_u32`, which is used: to "atomically" check if the blocked modes changed within the call of `pm_set_lowest` and to set all block modes to 0 in the power_off method (in the latest we are not in a hurry, one could set all to 0 in a loop as well).